File: apps/frontend/lib/models/ai_config.dart
import 'package:uuid/uuid.dart';

// AI configuration model
enum AIProvider {
  openai,
  claude,
  ollama,
}

// UUID helper
final _uuid = Uuid();

class AIConfig {
  final String id;
  final AIProvider provider;
  final String apiEndpoint;
  final String modelName;
  final int maxTokens;
  final bool isEnabled;
  final DateTime createdAt;
  final DateTime updatedAt;

  AIConfig({
    String? id,
    required this.provider,
    required this.apiEndpoint,
    required this.modelName,
    this.maxTokens = 1000,
    this.isEnabled = false,
    required this.createdAt,
    required this.updatedAt,
  }) : id = id ?? _uuid.v4();

  factory AIConfig.fromJson(Map<String, dynamic> json) {
    return AIConfig(
      id: json['id'] as String,
      provider: AIProvider.values.firstWhere(
        (e) => e.name == json['provider'],
        orElse: () => AIProvider.openai,
      ),
      apiEndpoint: json['api_endpoint'] as String,
      modelName: json['model_name'] as String,
      maxTokens: json['max_tokens'] as int? ?? 1000,
      isEnabled: json['is_enabled'] == 1,
      createdAt: DateTime.fromMillisecondsSinceEpoch(json['created_at'] * 1000),
      updatedAt: DateTime.fromMillisecondsSinceEpoch(json['updated_at'] * 1000),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'provider': provider.name,
      'api_endpoint': apiEndpoint,
      'model_name': modelName,
      'max_tokens': maxTokens,
      'is_enabled': isEnabled ? 1 : 0,
      'created_at': createdAt.millisecondsSinceEpoch ~/ 1000,
      'updated_at': updatedAt.millisecondsSinceEpoch ~/ 1000,
    };
  }
}

// SyncCredential model
class SyncCredential {
  final String id;
  final String endpoint;
  final String bucketName;
  final String? region;
  final bool isEnabled;
  final DateTime createdAt;
  final DateTime updatedAt;

  SyncCredential({
    String? id,
    required this.endpoint,
    required this.bucketName,
    this.region,
    this.isEnabled = false,
    required this.createdAt,
    required this.updatedAt,
  }) : id = id ?? _uuid.v4();

  factory SyncCredential.fromJson(Map<String, dynamic> json) {
    return SyncCredential(
      id: json['id'] as String,
      endpoint: json['endpoint'] as String,
      bucketName: json['bucket_name'] as String,
      region: json['region'] as String?,
      isEnabled: json['is_enabled'] == 1,
      createdAt: DateTime.fromMillisecondsSinceEpoch(json['created_at'] * 1000),
      updatedAt: DateTime.fromMillisecondsSinceEpoch(json['updated_at'] * 1000),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'endpoint': endpoint,
      'bucket_name': bucketName,
      'region': region,
      'is_enabled': isEnabled ? 1 : 0,
      'created_at': createdAt.millisecondsSinceEpoch ~/ 1000,
      'updated_at': updatedAt.millisecondsSinceEpoch ~/ 1000,
    };
  }
}

// ExportArchive model
class ExportArchive {
  final String id;
  final String filePath;
  final String checksum;
  final int sizeBytes;
  final int itemCount;
  final bool isEncrypted;
  final DateTime createdAt;

  ExportArchive({
    String? id,
    required this.filePath,
    required this.checksum,
    required this.sizeBytes,
    required this.itemCount,
    this.isEncrypted = true,
    required this.createdAt,
  }) : id = id ?? _uuid.v4();

  factory ExportArchive.fromJson(Map<String, dynamic> json) {
    return ExportArchive(
      id: json['id'] as String,
      filePath: json['file_path'] as String,
      checksum: json['checksum'] as String,
      sizeBytes: json['size_bytes'] as int,
      itemCount: json['item_count'] as int,
      isEncrypted: json['is_encrypted'] == 1,
      createdAt: DateTime.fromMillisecondsSinceEpoch(json['created_at'] * 1000),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'file_path': filePath,
      'checksum': checksum,
      'size_bytes': sizeBytes,
      'item_count': itemCount,
      'is_encrypted': isEncrypted ? 1 : 0,
      'created_at': createdAt.millisecondsSinceEpoch ~/ 1000,
    };
  }
}

File: packages/backend/internal/models/ai_config.go
// Package models provides data model definitions for MemoNexus Core.
package models

import (
	"time"

	"github.com/kimhsiao/memonexus/backend/internal/crypto"
)

// AIConfig holds encrypted AI service configuration.
// APIKeyEncrypted is never exposed in JSON responses.
type AIConfig struct {
	ID              UUID   `db:"id" json:"id"`
	Provider        string `db:"provider" json:"provider"` // openai, claude, ollama
	APIEndpoint     string `db:"api_endpoint" json:"api_endpoint"`
	APIKeyEncrypted string `db:"api_key_encrypted" json:"-"` // Never expose
	ModelName       string `db:"model_name" json:"model_name"`
	MaxTokens       int    `db:"max_tokens" json:"max_tokens"`
	IsEnabled       bool   `db:"is_enabled" json:"is_enabled"`
	CreatedAt       int64  `db:"created_at" json:"created_at"`
	UpdatedAt       int64  `db:"updated_at" json:"updated_at"`
}

// TableName returns the table name for AIConfig.
func (AIConfig) TableName() string {
	return "ai_config"
}

// CreatedAtTime returns the CreatedAt as time.Time.
func (a *AIConfig) CreatedAtTime() time.Time {
	return time.Unix(a.CreatedAt, 0)
}

// UpdatedAtTime returns the UpdatedAt as time.Time.
func (a *AIConfig) UpdatedAtTime() time.Time {
	return time.Unix(a.UpdatedAt, 0)
}

// =====================================================
// T226: API Key Encryption/Decryption (AES-256-GCM)
// =====================================================

// SetAPIKey encrypts and sets the API key using AES-256-GCM.
// T226: Encryption at rest for sensitive credentials.
func (a *AIConfig) SetAPIKey(apiKey, machineID string) error {
	encrypted, err := crypto.EncryptAPIKey(apiKey, machineID)
	if err != nil {
		return err
	}
	a.APIKeyEncrypted = encrypted
	return nil
}

// GetAPIKey decrypts and returns the API key.
// T226: Decryption of encrypted credentials at rest.
func (a *AIConfig) GetAPIKey(machineID string) (string, error) {
	if a.APIKeyEncrypted == "" {
		return "", nil
	}
	return crypto.DecryptAPIKey(a.APIKeyEncrypted, machineID)
}

// HasAPIKey returns true if an encrypted API key is stored.
func (a *AIConfig) HasAPIKey() bool {
	return a.APIKeyEncrypted != ""
}
